lib/SecretsCrypto/sf-secrets-manager.c

```diff
@@ -0,0 +1,830 @@
+			"id", id,
+			"invocation", invocation,
+			NULL);
+
+	(void)args;
+
+	g_object_add_weak_pointer(G_OBJECT(request), (gpointer *)&request);
+	g_signal_emit(manager,
+			_sf_secrets_manager_signals[SIGNAL_NEW_INTERACTION_REQUEST], 0,
+			request,
+			&result);
+	g_object_unref(request);
+
+	if ((request = g_object_ref(request))) {
+		g_object_remove_weak_pointer(G_OBJECT(request), (gpointer *)&request);
+		g_hash_table_insert(priv->requests, id, request);
+		g_object_weak_ref(G_OBJECT(request), (GWeakNotify)_sf_secrets_manager_remove_request, manager);
+		g_object_unref(request);
+	}
+
+	return TRUE;
+}
+
+static gboolean _sf_secrets_manager_continue_interaction(SfSecretsManager *manager,
+		GDBusMethodInvocation *invocation,
+		const gchar *request_id,
+		GVariant *request_params)
+{
+	SfSecretsManagerPrivate *priv = sf_secrets_manager_get_instance_private(manager);
+	SfSecretsInteractionRequest *request = g_hash_table_lookup(priv->requests, request_id);
+
+	(void)request_params;
+
+	if (!request)
+		return FALSE;
+
+	g_object_set(request,
+			"invocation", invocation,
+			NULL);
+	g_signal_emit_by_name(request, "continue");
+
+	return TRUE;
+}
+
+static gboolean _sf_secrets_manager_cancel_interaction(SfSecretsManager *manager,
+		GDBusMethodInvocation *invocation,
+		const gchar *request_id)
+{
+	SfSecretsManagerPrivate *priv = sf_secrets_manager_get_instance_private(manager);
+	SfSecretsInteractionRequest *request = g_hash_table_lookup(priv->requests, request_id);
+
+	if (!request)
+		return FALSE;
+
+	g_hash_table_remove(priv->requests, request_id);
+	g_object_weak_unref(G_OBJECT(request), (GWeakNotify)_sf_secrets_manager_remove_request, manager);
+	g_object_set(request,
+			"invocation", invocation,
+			NULL);
+	g_signal_emit_by_name(request, "cancel");
+
+	return TRUE;
+}
+
+static gboolean _sf_secrets_manager_finish_interaction(SfSecretsManager *manager,
+		GDBusMethodInvocation *invocation,
+		const gchar *request_id)
+{
+	SfSecretsManagerPrivate *priv = sf_secrets_manager_get_instance_private(manager);
+	SfSecretsInteractionRequest *request = g_hash_table_lookup(priv->requests, request_id);
+
+	if (!request)
+		return FALSE;
+
+	g_hash_table_remove(priv->requests, request_id);
+	g_object_weak_unref(G_OBJECT(request), (GWeakNotify)_sf_secrets_manager_remove_request, manager);
+	g_object_set(request,
+			"invocation", invocation,
+			NULL);
+	g_signal_emit_by_name(request, "finish");
+
+	return TRUE;
+}
+
+static void _sf_secrets_manager_set_property(GObject *object, guint property_id, const GValue *value, GParamSpec *spec)
+{
+	SfSecretsManager *manager = SF_SECRETS_MANAGER(object);
+	SfSecretsManagerPrivate *priv = sf_secrets_manager_get_instance_private(manager);
+
+	(void)spec;
+
+	switch (property_id) {
+		case PROP_APPLICATION_ID:
+			if (priv->application_id)
+				g_free(priv->application_id);
+			priv->application_id = g_value_dup_string(value);
+			break;
+		case PROP_USER_INTERACTION_MODE:
```

Comment:
I wonder how many C applications will need to support in-app interaction mode?  I mean, if they provide a UI to allow the user to perform password entry or so on, I would assume that the UI would be a Sailfish.Silica / QML-based UI; and if that's the case, they should probably just use the C++ Secrets API.

I'm sure there are specific use cases for this (e.g. minimal UI case, some daemon which just talks to USB-key plugin), so it's probably needed eventually, but IMO getting everything else working first is probably higher priority.

Reply:
I agree, this should've waited, I just wasn't familiar enough with the API yet when starting the implementation...

Reply:
Understandable, it's quite a complex domain.  If you ever have any questions about anything, just yell :-)

Reply:
In principle a C app could also do some sort of in-app authentication, maybe in CLI rather than GUI?

Reply:
Very true, and it does need to be done eventually for completeness, but I just think it's lower priority than e.g. getting the crypto API finished.